source/text.c

```diff
@@ -249,24 +283,242 @@ void C2D_DrawText(const C2D_Text* text, u32 flags, float x, float y, float z, fl
 	}
 	if (flags & C2D_WithColor)
 		color = va_arg(va, u32);
+	if (flags & C2D_WordWrap)
+		maxWidth = (float)va_arg(va, double); // Passed as float, but varargs promotes to double.
 
 	va_end(va);
 
 	C2Di_SetCircle(false);
 
-	for (cur = begin; cur != end; ++cur)
+	// Get information about lines
+	struct line {
+		u32 words;
+		u32 wordStart;
+	};
+	struct line lines[text->lines];
+	memset(lines, 0, sizeof(struct line) * text->lines);
+	for (cur = begin; cur != end; cur++)
+	{
+		if (cur->wordNo >= lines[cur->lineNo].words)
+			lines[cur->lineNo].words = cur->wordNo + 1;
+	}
+	for (u32 i = 1; i < text->lines; i++)
+	{
+		lines[i].wordStart = lines[i-1].wordStart + lines[i-1].words;
+	}
+
+	// Get information about words
+	struct word {
+		C2Di_Glyph* start;
+		C2Di_Glyph* end;
+		float wrapXOffset;
+		u32 newLineNumber;
+	};
+	struct word words[text->words];
```

Comment:
Same here - I think this should not be used nor calculated for every single text draw operation, only for those special modes that need it